packages/auth-core/src/core.ts

```diff
@@ -0,0 +1,291 @@
+  }
+
+  getToken(code: string, verifier: string): Promise<TokenData> {
+    return this._fetch("token", "get", {
+      code,
+      verifier,
+    }) as Promise<TokenData>;
+  }
+
+  async signinWithEmailPassword(email: string, password: string) {
+    const { challenge, verifier } = pkce.createVerifierChallengePair();
+    const { code } = (await this._fetch("authenticate", "post", undefined, {
+      provider: "builtin::local_emailpassword",
+      challenge,
+      email,
+      password,
+    })) as { code: string };
+    return this.getToken(code, verifier);
+  }
+
+  async signupWithEmailPassword(
+    email: string,
+    password: string,
+    verifyUrl: string
+  ): Promise<
+    | { status: "complete"; tokenData: TokenData }
+    | { status: "verificationRequired"; verifier: string }
+  > {
+    const { challenge, verifier } = pkce.createVerifierChallengePair();
+    const result = (await this._fetch("register", "post", undefined, {
+      provider: "builtin::local_emailpassword",
+      challenge,
+      email,
+      password,
+      verify_url: verifyUrl,
+    })) as { code: string } | { verification_email_sent_at: string };
+    if ("code" in result) {
+      return {
+        status: "complete",
+        tokenData: await this.getToken(result.code, verifier),
+      };
+    } else {
+      return { status: "verificationRequired", verifier };
+    }
+  }
+
+  async verifyEmailPasswordSignup(verificationToken: string, verifier: string) {
+    const { code } = (await this._fetch("verify", "post", undefined, {
+      provider: "builtin::local_emailpassword",
+      verification_token: verificationToken,
+    })) as { code: string };
+    return this.getToken(code, verifier);
+  }
+
+  async resendVerificationEmail(verificationToken: string) {
+    await this._fetch("resend-verification-email", "post", undefined, {
+      provider: "builtin::local_emailpassword",
+      verification_token: verificationToken,
+    });
+  }
+
+  async sendPasswordResetEmail(email: string, resetUrl: string) {
+    return this._fetch("send-reset-email", "post", undefined, {
+      provider: "builtin::local_emailpassword",
+      email,
+      reset_url: resetUrl,
+    }) as Promise<{ email_sent: string }>;
+  }
+
+  static checkPasswordResetTokenValid(resetToken: string) {
+    try {
+      const payload = jwtDecode(resetToken);
+      if (
+        typeof payload !== "object" ||
+        payload == null ||
+        !("exp" in payload) ||
+        typeof payload.exp !== "number"
+      ) {
+        return false;
+      }
+      return payload.exp < Date.now();
+    } catch {
+      return false;
+    }
+  }
+
+  async resetPasswordWithResetToken(resetToken: string, password: string) {
+    return this._fetch("reset-password", "post", undefined, {
+      provider: "builtin::local_emailpassword",
+      reset_token: resetToken,
+      password,
+    }) as Promise<TokenData>;
+  }
+
+  async getProvidersInfo() {
+    // TODO: cache this data somehow?
```

Comment:
You can memoize easily on a [private class property](https://www.typescriptlang.org/play?#code/C4TwDgpgBAShDOBXANsKBeKBvAUFfUAJgIbDEBcUARgPY3ITEB2UOAvgNxQD03UokfjSgAnCGDHwITNMAAW0AMbFFCwkVLEcORcmLx4UAILY8BXlAkBLAG6loNMMCs0mxZFABmViMkJn8AGJlVQgAfko4JFQObQIoAHMIYAARTQAKAEpTePirTyh0+St4ADpglQVs3FzaxVd4eghS5BoE9IAiEIUoOStgDszY2tyxYEQRFmKyitDhkbY4kfqmRoYWts7u6ABbEvhB+dqLHZpFAGsvZND1EjIA3JX4NEkUNEwsO4p+EUQITge8Qs3mQHm2gII03K2ww2C+lGAv3+sQh+DGEymfRm2yOUDYrEWOgaaGIsKYEAA7sYstpiKUkqkMtkcHSGWkyDTWcl2cQaUA)

Reply:
Yes, while the caching bit is easy, this note was more of a reminder to add it once we figure out a way to invalidate the cached provider info when the config is updated.